File: typescript/benchmarks/benchmark.mjs
import { readFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { performance } from "node:perf_hooks";
import { fileURLToPath } from "node:url";

import {
  normalizeToIR,
  projectCanonical,
  projectHypabolic,
  projectLetta,
  projectMinimalJsonl,
  projectOpenAI,
  serializeProjection,
} from "@hypabolic/trajectory";
import { projectOpenTelemetry } from "@hypabolic/trajectory-otel";

const workspace = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const transcriptBytes = await readFile(
  resolve(workspace, "../conformance/cases/pi/unicode-boundaries/input.jsonl"),
);
const request = { source: "pi", transcriptBytes };
const iterations = 250;

for (let index = 0; index < 10; index++) normalizeToIR(request);
const heapBefore = process.memoryUsage().heapUsed;
const started = performance.now();
let trajectory;
for (let index = 0; index < iterations; index++) {
  trajectory = normalizeToIR(request);
}
const elapsed = performance.now() - started;
const heapDelta = process.memoryUsage().heapUsed - heapBefore;
const outputs = {
  "letta-trajectory-v1": serializeProjection(projectLetta(trajectory)),
  "letta-canonical-v1": serializeProjection(projectCanonical(trajectory)),
  "hypabolic-trajectory-v1": serializeProjection(projectHypabolic(trajectory)),
  "openai-chat-messages": serializeProjection(projectOpenAI(trajectory)),
  "jsonl-minimal": projectMinimalJsonl(trajectory),
  "otel-genai-spans-v1": serializeProjection(projectOpenTelemetry(trajectory)),
};

console.log(JSON.stringify({
  runtime: "typescript",
  iterations,
  elapsed_ms: elapsed,
  operations_per_second: iterations / (elapsed / 1_000),
  heap_delta_bytes: heapDelta,
  input_bytes: transcriptBytes.byteLength,
  records: trajectory.records.length,
  output_bytes: Object.fromEntries(
    Object.entries(outputs).map(([name, output]) => [
      name,
      Buffer.byteLength(output),
    ]),
  ),
}));

File: rust/crates/hypabolic-trajectory/examples/benchmark.rs
//! Dependency-free representative normalization and output-size benchmark.

use std::collections::BTreeMap;
use std::hint::black_box;
use std::time::Instant;

use hypabolic_trajectory::{
    NormalizeRequest, PiSourceAdapter, SourceAdapter, project_canonical, project_hypabolic,
    project_letta, project_minimal_jsonl, project_openai, project_opentelemetry,
};
use serde_json::json;

fn main() {
    const ITERATIONS: u32 = 250;
    let transcript =
        include_bytes!("../../../../conformance/cases/pi/unicode-boundaries/input.jsonl");
    let request = NormalizeRequest {
        transcript,
        source_context: Default::default(),
        options: Default::default(),
    };
    for _ in 0..10 {
        black_box(PiSourceAdapter.normalize(request).expect("warmup"));
    }
    let started = Instant::now();
    let mut trajectory = None;
    for _ in 0..ITERATIONS {
        trajectory = Some(
            PiSourceAdapter
                .normalize(request)
                .expect("benchmark normalization"),
        );
    }
    let elapsed = started.elapsed();
    let trajectory = trajectory.expect("benchmark has iterations");
    let mut outputs = BTreeMap::new();
    outputs.insert(
        "hypabolic-trajectory-v1",
        project_hypabolic(&trajectory)
            .expect("Hypabolic output")
            .len(),
    );
    outputs.insert(
        "jsonl-minimal",
        project_minimal_jsonl(&trajectory)
            .expect("minimal output")
            .len(),
    );
    outputs.insert(
        "letta-canonical-v1",
        project_canonical(&trajectory)
            .expect("canonical output")
            .len(),
    );
    outputs.insert(
        "letta-trajectory-v1",
        project_letta(&trajectory).expect("Letta output").len(),
    );
    outputs.insert(
        "openai-chat-messages",
        project_openai(&trajectory).expect("OpenAI output").len(),
    );
    outputs.insert(
        "otel-genai-spans-v1",
        project_opentelemetry(&trajectory)
            .expect("OpenTelemetry output")
            .len(),
    );
    println!(
        "{}",
        serde_json::to_string(&json!({
            "runtime": "rust",
            "iterations": ITERATIONS,
            "elapsed_ms": elapsed.as_secs_f64() * 1_000.0,
            "operations_per_second": f64::from(ITERATIONS) / elapsed.as_secs_f64(),
            "input_bytes": transcript.len(),
            "records": trajectory.records.len(),
            "output_bytes": outputs,
            "allocation_note": "stable Rust exposes no safe process-local allocation counter",
        }))
        .expect("benchmark result serializes")
    );
}
